doc/adr004-effect-system.md

```diff
@@ -38,150 +36,292 @@ simplicity.
+this is not straightforward.
+
+### Entities
+
+Effects act upon entities, which are typically lists of state variables.
+However, we also need to represent entities as a variable that can later be
+replaced with the actual state variables and combinations (unions) of those.
+This is necessary for defining effect signatures for operators.
+
+Consider an operator such as `+` (`iadd`), which takes two arguments, each with
+its own effect, and returns a value whose effect is the combination of the
+effects of the arguments. We could write the signature of `iadd` as:
+
+```
+iadd: (Read[r1] & Temporal[t1]) => (Read[r2] & Temporal[t2]) => Read[r1, r2] & Temporal[t1, t2]
+```
+
+Here, `r1`, `r2`, `t1` and `t2` are entity variables, which can be substituted
+with the actual state variables during inference.
+
+A concrete entity with no state variables and an union of zero entities are both
+considered empty entities, and we say them nullify the effect actin upon them.
+For example, replacing `r` with an empty entity in `Read[r]` results in `Read[]`
+which is equivalent to `Pure`.
+
+```typescript
+/*
+ * The Entity an effect acts upon. Either a list of state variables, a
+ * variable or a combination of other entities.
+ */
+export type Entity =
+  /* A list of state variables */
+  | { kind: 'concrete', stateVariables: StateVariable[] }
+  /* A variable representing some entity */
+  | { kind: 'variable', name: string, reference?: bigint }
+  /* A combination of entities */
+  | { kind: 'union', entities: Entity[] }
+```
+
+A state variable is just a name that is exactly the name of a state variable in
+the spec. It also carries an optional reference to the expression on which the
+state variable occurred, which is used for error reporting.
+
+```typescript
+/* A state variable */
+export interface StateVariable {
+  /* The variable name */
+  name: string,
+  /* The id of the expression on which the state variable ocurred */
+  reference: bigint
+}
 ```
 
-### Normal form
+### Effects
 
-A normalized effect takes only one of two forms:
+An effect can also have three different forms: concrete, arrow and variable.
 
+```typescript
+/*
+ * The effect of a Quint expression, regarding which state entities are read
+ * and/or updated
+ */
+export type Effect =
+  | ConcreteEffect
+  | ArrowEffect
+  | EffectVariable
 ```
-Effects E ::= Read[vars] & Update[vars] | (E0, ..., EN) => E
+
+The most basic representation of effects is a concrete effect, and it is
+structured as a list of effect components. An effect component has a kind and an
+entity. The kind can be `read`, `update` or `temporal`, and the entity is the
+entity the effect acts upon.
+
+```typescript
+/* Kinds for concrete effect components */
+export type ComponentKind = 'read' | 'update' | 'temporal'
+
+/* A concrete effect component, specifying a kind and the entity it acts upon */
+export interface EffectComponent { kind: ComponentKind, entity: Entity }
```

Comment:
I'm not sure why we need the `EffectComponent` now. Iiuc, it's just a product of an effect kind and entity, so maybe using a tuple would suffice and reduce the number of abstract nouns we have to track by one? Just thinking out loud... :)

Reply:
We have to define a name for the field in `ConcreteEffect` that carries this, which is now `components`. So I don't think we can reduce the number, and this might help clarify it.